Added TweenSequence, TweenSequenceItem, ConstantTween

HansMuller opened the pull request:
Make it possible to define an animation by stringing together a sequence of tweens.

HansMuller committed:
TweenSequence, TweenSequenceItem, ConstantTween

Hixie commented:
CurveTween is an `Animatable`, which is the superclass of `Curve`. I think if `TweenSequenceItem` were to take an `Animatable` instead of a `Curve`, we'd be able to get rid of the `curve` property and use composition instead.

Hixie commented on packages/flutter/lib/src/animation/tween_sequence.dart:
```diff
@@ -0,0 +1,128 @@
+  /// [TweenSequence] animation's duration when [tween] will be used.
+  ///
+  /// The percentage for an individual item is the item's weight divided
+  /// by the sum of all of the items' weights.
```
oh i see that's in the constructor docs already, never mind

Hixie commented on packages/flutter/lib/src/animation/tween_sequence.dart:
````diff
@@ -0,0 +1,128 @@
+///   ],
+/// ).animate(myAnimationController);
+///```
+class TweenSequence<T> extends Animatable<T> {
````
we should mention that actually creating one of these is expensive (so that people don't just create them in build methods and throw them away every frame)

Hixie commented on packages/flutter/lib/src/animation/tween_sequence.dart:
```diff
@@ -0,0 +1,128 @@
+    for (TweenSequenceItem<T> item in _items)
+      totalWeight += item.weight;
+
+    double start = 0.0;
```
assert totalWeight > 0

Hixie commented on packages/flutter/lib/src/animation/tween_sequence.dart:
```diff
@@ -0,0 +1,128 @@
+    for (int i = 0; i < _items.length; i++) {
+      final double end = start + _items[i].weight / totalWeight;
+      _intervals.add(new _Interval(start, end));
+      start = end;
```
I'm a little concerned about rounding errors here. they'll add up as you add each item, and you won't reach precisely 1.0. Maybe keep track of the sum of all the weights so far and use that to compute the end, rather than adding start? Though that still won't guarantee reaching 1.0...

HansMuller replied:
Good point. I've forced the end of the last item to be 1.0.

Hixie commented:
![LGTM](http://junkyard.damowmow.com/577)

Hixie commented:
This is great, I love it.

HansMuller committed:
Updated per review feedback

HansMuller committed:
fixed a typo

HansMuller commented:
Although I left the name "TweenSequence", since that's likely to be the only use case, I changed the type of `TweeSequenceElement<T>.tween` to `Animatable<T>`, and removed the curve parameter. 

To evaluate an Animatable at `t` I used the same approach as the implementation of `Animatable.chain`:
```
tween.evaluate(new AlwaysStoppedAnimation<double>(t));
```

HansMuller committed:
Updated per review feedback

namiwang commented:
Guys, maybe also update `https://flutter.io/animations/staggered-animations/` to reflect this update?
